### PR TITLE
Hawkular-464 Deployment Failure Event Alerting

### PR DESCRIPTION
**Pull request by jshaughn**
Add alert generation for failed deployments.
- New src-deps versions for component enhancements to support this
  - with mazz, enhanced bus-commons with better classloading/deserialization support for MDBs
  - with mazz, enhanced command-gateway to support multiple commands on ws responses, now can notify UI and put a response on the bus
  - enhanced alerts to consume ws deploy app response messages on bus and generate events
- Added new trigger definition on app-servers to fire on a failed deployment event
  - note, also storing successful deployment events for future use/display
- Added new UI support for 'Event' trigger-type and DEPLOYMENT_FAIL alert-type

Also:
- fixed time-window bug in getAlerts query for some trigger types
- make bus-commons dep provided (by nest) for pinger
- make bus-commons dep provided (by nest) for avail-creator

### CODE REVIEW

File: console/src/main/scripts/plugins/metrics/ts/app-details/appServerDatasourcesDetails.ts [hunk 1]
```diff
@@ -240,6 +224,27 @@ module HawkularMetrics {
       this.getDrivers();
     }
 
+    private getAlerts(res:any):void {
```
Comment (mtho11) [inline console/src/main/scripts/plugins/metrics/ts/app-details/appServerDatasourcesDetails.ts]: Would be nice to have the type here.

File: console/src/main/scripts/plugins/metrics/ts/app-details/appServerDatasourcesDetails.ts [hunk 2]
```diff
@@ -240,6 +224,27 @@ module HawkularMetrics {
...
+      let dsArray:IAlert[];
+      let promise = this.HawkularAlertsManager.queryAlerts({
+        statuses: 'OPEN',
+        tags: 'resourceId|' + res.id,
+        startTime: this.startTimeStamp,
+        endTime: this.endTimeStamp
+      }).then((data)=> {
```
Comment (mtho11) [inline console/src/main/scripts/plugins/metrics/ts/app-details/appServerDatasourcesDetails.ts]: The type would be good here too as anyone outside of alerts wouldn't know what kind of data is in 'data'. The types let our IDE's autocomplete this stuff.

File: console/src/main/scripts/plugins/metrics/ts/app-details/appServerDatasourcesDetails.ts [hunk 3]
```diff
@@ -240,6 +224,27 @@ module HawkularMetrics {
...
+        _.forEach(data.alertList, (item) => {
+          item['alertType'] = item.context.alertType;
```
Comment (mtho11) [inline console/src/main/scripts/plugins/metrics/ts/app-details/appServerDatasourcesDetails.ts]: It is preferable to type the item(IAlert) then use item.alertType instead of item['alertType'] which is not typed. This just helps to cuts down on typos and stupid little errors in the long run.

File: console/src/main/scripts/plugins/metrics/ts/app-details/appServerDatasourcesDetails.ts (hunk 1 above)
Comment (jshaughn) [inline console/src/main/scripts/plugins/metrics/ts/app-details/appServerDatasourcesDetails.ts]: Yes, this is actually not new, it's an artifact of the original code that I've been working on.  Same for the 'data' below.  I can look to see if I can improve it...

File: console/src/main/scripts/plugins/metrics/ts/app-details/appServerDeploymentsDetails.ts [hunk 4]
```diff
@@ -57,7 +58,8 @@ module HawkularMetrics {
       $scope.vm = this;
       HawkularOps.init(this.NotificationsService);
 
-      this.startTimeStamp = +moment().subtract(1, 'hours');
+      this.resourceId = this.$routeParams.resourceId;
+      this.startTimeStamp = +moment().subtract(($routeParams.timeOffset || 3600000), 'milliseconds');
```
Comment (mtho11) [inline console/src/main/scripts/plugins/metrics/ts/app-details/appServerDeploymentsDetails.ts]: We need to handle this default time period a bit better, so that if we change it in one place it gets changed everywhere. Nothing you need to do here just making a note of this. Ultimately, it should probably be user-definable anyway.

File: console/src/main/scripts/plugins/metrics/ts/app-details/appServerJvmDetails.ts [hunk 5]
```diff
@@ -126,36 +146,16 @@ module HawkularMetrics {
...
-      this.getAlerts(this.$routeParams.resourceId, this.startTimeStamp, this.endTimeStamp);
-    }
-
     public getJvmData():void {
-      this.endTimeStamp = this.$routeParams.endTime || +moment();
-      this.startTimeStamp = this.endTimeStamp - (this.$routeParams.timeOffset || 3600000);
-
       this.MetricsService.retrieveGaugeMetrics(this.$rootScope.currentPersona.id,
-        'MI~R~[' + this.$routeParams.resourceId + '~~]~MT~WildFly Memory Metrics~Heap Used',
+        'MI~R~[' + this.resourceId + '~~]~MT~WildFly Memory Metrics~Heap Used',
```
Comment (mtho11) [inline console/src/main/scripts/plugins/metrics/ts/app-details/appServerJvmDetails.ts]: Why the change away from $routeParams all screens should be driven from the $routeParams to make our screens stateless and RESTful.
The url to a screen is like an API, one should be able to change the resourceId (or any other params in the uri) and the screen should change to reflect the new uri.

File: console/src/main/scripts/plugins/metrics/ts/metricsAppServerDetails.ts [hunk 6]
```diff
@@ -541,11 +541,64 @@ module HawkularMetrics {
...
+        .existTrigger(this.resourceId + '_failed_deployment').then(() => {
+          this.$log.debug('Failed Deployment trigger exists, nothing to do');
+        }, () => {
+          // Failed Deployment trigger doesn't exist, need to create one
+
+          let triggerId:string = this.resourceId + '_failed_deployment';
+          let resourceId:string = triggerId.slice(0, -18);
+          let dataId:string = this.resourceId + '_DeployApplicationResponse';
+
+          let fullTrigger = {
```
Comment (mtho11) [inline console/src/main/scripts/plugins/metrics/ts/metricsAppServerDetails.ts]: Add typing around fullTrigger. 
Replace with 'let fullTrigger:IAlertTrigger = {'

File: console/src/main/scripts/plugins/metrics/ts/alertsCenterTriggerList.ts [hunk 7]
```diff
@@ -190,6 +190,8 @@ module HawkularMetrics {
 
       if ('Availability' === trigger.context.triggerType) {
         route = '/hawkular-ui/alerts-center-triggers/availability/' + encodedId;
+      } else if ('Event' === trigger.context.triggerType) {
```
Comment (mtho11) [inline console/src/main/scripts/plugins/metrics/ts/alertsCenterTriggerList.ts]: Switch might be easier to maintain?

File: console/src/main/scripts/plugins/metrics/ts/app-details/appServerJvmDetails.ts (hunk 5 above)
Comment (jshaughn) [inline console/src/main/scripts/plugins/metrics/ts/app-details/appServerJvmDetails.ts]: In the constructor we have:

  this.resourceId = this.$routeParams.resourceId;

I just used the local var instead of going through routParams everywhere. Is that not OK?

Comment (mtho11) [inline console/src/main/scripts/plugins/metrics/ts/app-details/appServerJvmDetails.ts]: No, otherwise this is bypassing the RESTful params.

Comment (jshaughn) [inline console/src/main/scripts/plugins/metrics/ts/app-details/appServerJvmDetails.ts]: I'll change it back but I don't really get it, I don't know then what the point is of a variable scoped at the class level.  Are you saying we want to support hand-editing of the url without actually having the class get re-constructed?

Comment (mtho11) [inline console/src/main/scripts/plugins/metrics/ts/app-details/appServerJvmDetails.ts]: @jshaughn sorry you are correct about:'this.resourceId = this.$routeParams.resourceId;' being equivalent. When you are just looking through the diffs and you see a change from *$routeParams.resourceId* to *resourceId* it looks like a bad change ;)

Comment (mtho11) [inline console/src/main/scripts/plugins/metrics/ts/app-details/appServerJvmDetails.ts]: For more self describing code: the *$routeParams.resourceId* is clear where it came from whereas *resourceId* is not clear that it came from the url (you have to look into where it was assigned somewhere else in the code).

**Comment by mtho11** [comment]
Merged #675
